### ISSUE TITLE
CRSF input for Trainer mode

### ISSUE THREAD

**Issue by sk8board** [EdgeTX/edgetx#1607]
Currently, the Trainer mode only allows PPM or SBUS input. I am requesting CRSF input for Trainer mode.

Here is why.

I place a receiver in the JR module bay and connect the receiver to the Batt, Ground, and HeartBeat terminals. If the receiver will output PPM or SBUS, then this will work for binding a student transmitter to the receiver in the JR module bay. Now wireless trainer mode can be used.

With internal ELRS modules becoming popular, now transmitters will have to use only ELRS receivers. The problem is ELRS receivers do not output PPM or SBUS. Therefore, there is a need for trainer mode to support CRSF protocol.

**Comment by rotorman** [comment]
Luckily the channel data in CRSF is encoded 1-to-1 as with S.BUS. Here as a helper, a short CRSF parsing code by kobata, where he connected an HappyModel EP2 receiver via MCU to USB to act as a joystick: https://qiita.com/kobatan/items/40728fbb625057d9f42b

**Comment by wimalopaan** [comment]
Does anyone have a full protocol spec at hand? For this the control-part (no sensor reply) would be sufficient. 

What would be the best test-setup for someone not owning CRSF hardware? Are the PC-based generators e.g. to use with some USB/serial converter?

**Comment by rotorman** [comment]
Correct, you could just wire up the EP2 (CRSF) receiver TX line and leave the RX unconnected. Here a capture from the EP2 TX line (snippet):

```
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21 
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
C8 0C 14 35 00 64 0C 00 07 00 00 00 00 2E
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21 
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
```

**Comment by wimalopaan** [comment]
> Correct, you could just wire up the EP2 (CRSF) receiver TX line and leave the RX unconnected. Here a capture from the EP2 TX line (snippet):
> 
> ```
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21 
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
> C8 0C 14 35 00 64 0C 00 07 00 00 00 00 2E
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21 
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
> C8 18 16 EC 03 9F 49 99 F3 0B F0 81 0F 7C E0 03 1F F8 28 08 00 00 44 1C D7 21
> ```

Cool, thanks for the data: will use that as replay ...

**Comment by wimalopaan** [comment]
If I would like to get real HW, what would you suggest as RF-Module and RX? Available, simple and cheap...

**Comment by wimalopaan** [comment]
The other protocols like SBUS, IBUS oder SumD/V3 use a fixed known baudrate. As I read CRSF can use different baudrates. But in the hardware-tab there is no more a baudrate selector.

What would be your suggestion how to implement that (at first):

1) use a fixed baudrate
2) use a user selectable baudrate (we need again a baudrate chooser)
3) autobaud (this may be only a try-n-error strategy by simple trying some well known baudrates.

**Comment by rotorman** [comment]
Isn't the receiver side always 400k ?

**Comment by wimalopaan** [comment]
In #1462  now we have

* sbus (tested)
* ibus (tested)
* crsf (not tested)
* sumd (not implemented)

**Comment by wimalopaan** [comment]
For SUMDV3 I intend todo the following:

* first 16 channels are available as trainer input
* the binary switches 1-64 are mapped onto a LS iff the LS is of sticky type with no(!) set or reset condition.

So, if one wants to use the binary switches that come in via SUMDV3 trainer, she has to setup the wanted LS as above.

**Comment by pfeerick** [comment]
I'll grab another ELRS RX out later to try the CRSF support...

**Comment by wimalopaan** [comment]
> I'll grab another ELRS RX out later to try the CRSF support...

I do not guarantee for anything ;-)

**Comment by pfeerick** [comment]
It seems to be working fine... only checked with the first four channels so far though :)

**Comment by wimalopaan** [comment]
> It seems to be working fine... only checked with the first four channels so far though :)

Sounds good, thanks for testing.

**Comment by sk8board** [comment]
Thank you for working on this. Do you know if this enhancement is targeted for a release?

**Comment by pfeerick** [comment]
Should be in 2.8.

**Comment by sk8board** [comment]
@pfeerick @raphaelcoeffic 

> Should be in 2.8.

Is CRSF input for Trainer mode still planned for 2.8? If so, can I test this with 2.8.0-rc3?

**Comment by gagarinlg** [comment]
It is not going to be in 2.8.when you check PR #1462, you will see, that it needs to be rebased I clouding a bunch of manual work for translations.

**Comment by pfeerick** [comment]
So #1861 supercedes and includes #1462?

**Comment by wimalopaan** [comment]
Yes, it does.
It includes SBus/IBus/SumDV1-3 trainer input (up to 32 channels if EXTENDED_TRAINER is set) and FrSky-Telemetry-Input.

**Comment by raphaelcoeffic** [comment]
> It includes SBus/IBus/SumDV1-3 trainer input (up to 32 channels if EXTENDED_TRAINER is set) and FrSky-Telemetry-Input.

Kind of a "pot pourri", we'll need to sort that out and extract features one by one.

**Comment by wimalopaan** [comment]
Yes, it is. Actually I don't have time to sort that out by myself ... sorry.
But please don't hesitate to cherry pick the parts.

**Comment by pfeerick** [comment]
Lol... have no fear, the skeleton will be picked clean, and credit given where credit is due ;)
